web/packages/teleterm/src/mainProcess/agentDownloader/agentDownloader.test.ts

```diff
@@ -0,0 +1,176 @@
+
+import childProcess from 'node:child_process';
+import { PassThrough } from 'node:stream';
+import fs from 'node:fs';
+import zlib from 'node:zlib';
+
+import tarFs from 'tar-fs';
+
+import Logger, { NullService } from 'teleterm/logger';
+
+import { makeRuntimeSettings } from '../fixtures/mocks';
+
+import { downloadAgent } from './agentDownloader';
+
+import type { IFileDownloader } from './fileDownloader';
+
+jest.mock('node:child_process');
+jest.mock('node:fs');
+jest.mock('node:zlib');
+jest.mock('tar-fs');
+
+const LATEST_TELEPORT_VERSIONS_MOCK = [
+  {
+    releaseId: 'teleport@13.2.0',
+    product: 'teleport',
+    version: '13.2.0',
+    notesMd: '',
+    status: 'published',
+    assets: [],
+  },
+  {
+    releaseId: 'teleport@12.1.4',
+    product: 'teleport',
+    version: '12.1.4',
+    notesMd: '',
+    status: 'published',
+    assets: [],
+  },
+];
+
+beforeAll(() => {
+  Logger.init(new NullService());
+  // (Cannot spy the fetch property because it is not a function; undefined given instead)
+  // eslint-disable-next-line jest/prefer-spy-on
+  global.fetch = jest.fn().mockImplementation(() =>
+    Promise.resolve({
+      ok: true,
+      json: () => Promise.resolve(LATEST_TELEPORT_VERSIONS_MOCK),
+    })
+  );
+});
+
+beforeEach(() => {
+  jest.resetModules();
+  process.env = {};
+});
+
+const testCases = [
+  {
+    name: 'Should not download agent when Connect version and read version are identical',
+    connectVersion: '13.0.0',
+    versionFromCache: '13.0.0',
+    env: {},
+    shouldDownloadBinary: false,
+  },
+  {
+    name: 'Should download agent when Connect version and version from cache are different',
+    connectVersion: '13.0.0',
+    versionFromCache: '12.0.0',
+    env: {},
+    shouldDownloadBinary: 'teleport-v13.0.0-darwin-arm64-bin.tar.gz',
+  },
+  {
+    name: 'Should download agent when version from cache is missing',
+    connectVersion: '13.0.0',
+    versionFromCache: undefined,
+    env: {},
+    shouldDownloadBinary: 'teleport-v13.0.0-darwin-arm64-bin.tar.gz',
+  },
+  {
+    name: 'Should download the latest available agent version when Connect version is 1.0.0-dev',
+    connectVersion: '1.0.0-dev',
+    versionFromCache: undefined,
+    env: {},
+    shouldDownloadBinary: 'teleport-v13.2.0-darwin-arm64-bin.tar.gz',
+  },
+  {
+    name: 'Should download agent version from env var when Connect is 1.0.0-dev and env var is provided',
+    connectVersion: '1.0.0-dev',
+    versionFromCache: undefined,
+    env: {
+      CONNECT_CMC_AGENT_VERSION: '12.1.0',
+    },
+    shouldDownloadBinary: 'teleport-v12.1.0-darwin-arm64-bin.tar.gz',
+  },
+];
+
+test.each(testCases)(
+  '$name',
+  async ({ connectVersion, env, versionFromCache, shouldDownloadBinary }) => {
+    const runtimeSettings = makeRuntimeSettings({
+      tempDataDir: '/home/tmp',
+      agentBinaryPath: '/home/teleport/teleport',
+      sessionDataDir: '/home/cache',
+      appVersion: connectVersion,
+    });
+    process.env = env;
```

Comment:
As far as I know, this is not safe because what if another test also sets `process.env` at the same time? Instead of reading `process.env` in `downloadAgent` directly, you can pass it as an argument so that you don't have to interact with `process.env` in tests.

Reply:
Yeah, it can have some bad effects. Changed to arguments.